usaspending_api/disaster/v2/views/disaster_base.py

```diff
@@ -120,12 +120,14 @@ def all_closed_defc_submissions(self):
                 sub.fiscal_year == self.reporting_period_min_year
                 and sub.fiscal_period >= self.reporting_period_min_month
             ) or sub.fiscal_year > self.reporting_period_min_year:
-                q |= (
-                    Q(submission__reporting_fiscal_year=sub.fiscal_year)
-                    & Q(submission__quarter_format_flag=sub.is_quarter)
-                    & Q(submission__reporting_fiscal_period__lte=sub.fiscal_period)
+                q |= Q(
+                    submission__in=SubmissionAttributes.objects.filter(
+                        reporting_fiscal_year=sub.fiscal_year,
+                        quarter_format_flag=sub.is_quarter,
+                        reporting_fiscal_period__lte=sub.fiscal_period,
+                    )
                 )
-        return q & Q(submission__reporting_period_start__gte=str(self.reporting_period_min_date))
+        return q & ~Q(submission__toptier_code="nonsensical string")
```

Comment:
Currently a hackish solution for demonstration purposes. For some reason, simply having any inner join on the postgres query dramatically improves performance, so I've added this condition that will always pass that needs the submission table.

Reply:
> simply having any inner join on the postgres query dramatically improves performance

For performance, inner joins are preferred over outer joins when the data allows for it. If Django is converting the ORM to SQL with outer joins, it can typically be switched to an inner join if a `__isnull` filter is provided

Reply:
There don't appear to be any outer joins before or after this modification. Instead the difference is that in the origonal query:

SELECT
   "financial_accounts_by_awards"."award_id" 
FROM
   "financial_accounts_by_awards" 
   INNER JOIN
      "awards" 
      ON ("financial_accounts_by_awards"."award_id" = "awards"."id") 

submission_attributes is skipped (since there's no part of the where clause dependent on it), whereas when modified the joins expand to:

SELECT
   "financial_accounts_by_awards"."award_id" 
FROM
   "financial_accounts_by_awards" 
   INNER JOIN
      "submission_attributes" 
      ON ("financial_accounts_by_awards"."submission_id" = "submission_attributes"."submission_id") 
   INNER JOIN
      "awards" 
      ON ("financial_accounts_by_awards"."award_id" = "awards"."id") 


I'm still  trying to figure out why this would help performance instead of hurting, but the results of performance testing speak for themselves.

Reply:
Update: It seems the critical difference is that with the join added, a bitmap heap scan:

{{Bitmap Heap Scan on financial_accounts_by_awards  (cost=10043.92..13195.98 rows=144 width=8)
              Recheck Cond: ((submission_id = submission_attributes.submission_id) AND ((disaster_emergency_fund_code)::text = ANY ('{L,M,N,O,P}'::text[])))
              Filter: ((obligations_incurred_total_by_award_cpe > '0'::numeric) OR (obligations_incurred_total_by_award_cpe < '0'::numeric) OR (gross_outlays_delivered_orders_paid_total_cpe > '0'::numeric) OR (gross_outlays_delivered_orders_paid_total_cpe < '0'::numeric))}}

Is converted to an index scan. Still investigating why.